### PR TITLE
Prankster Gen 7

### PR DESCRIPTION
**Pull request by ghoulslash**
## Description
Dark types block priority-boosted status moves used by prankster-mons in gen7. Added a config option

Here's a GIF: poochyena has prankster
prankster

### CODE REVIEW

File: include/constants/battle_config.h [hunk 1]
```diff
@@ -122,6 +122,7 @@
 #define B_FLASH_FIRE_FROZEN         GEN_6 // In Gen5+, Flash Fire can trigger even when frozen, when it couldn't before.
 #define B_SYNCHRONIZE_NATURE        GEN_6 // In Gen8+, if the Pokémon with Synchronize is leading the party, it's 100% guaranteed that wild Pokémon will have the same ability, as opposed to 50% previously.
 #define B_UPDATED_INTIMIDATE        GEN_8 // In Gen8, Intimidate doesn't work on opponents with the Inner Focus, Scrappy, Own Tempo or Oblivious abilities.
+#define B_PRANKSTER                 GEN_7 // In Gen7, Prankster-elevated status moves do not affect Dark type Pokemon
```
Comment (AsparagusEduardo) [inline include/constants/battle_config.h:125]:
```suggestion
#define B_PRANKSTER_DARK_TYPES      GEN_7 // In Gen7, Prankster-elevated status moves do not affect Dark type Pokémon.
```

File: src/battle_util.c [hunk 2]
```diff
@@ -3185,6 +3186,24 @@ u8 AtkCanceller_UnableToUseMove(void)
             }
             gBattleStruct->atkCancellerTracker++;
             break;
+        case CANCELLER_PRANKSTER:
+            #if B_PRANKSTER >= GEN_7
```
Comment (AsparagusEduardo) [inline src/battle_util.c:3190]:
```suggestion
            #if B_PRANKSTER_DARK_TYPES >= GEN_7
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/battle_util.c [hunk 3]
```diff
@@ -3185,6 +3186,24 @@ u8 AtkCanceller_UnableToUseMove(void)
...
+                {
+                    if (!(gBattleTypeFlags & BATTLE_TYPE_DOUBLE) || !(gBattleMoves[gCurrentMove].target & (MOVE_TARGET_BOTH | MOVE_TARGET_FOES_AND_ALLY)))
+                        CancelMultiTurnMoves(gBattlerAttacker); // don't cancel moves that can hit two targets bc one target might not be protected
+                    gBattleScripting.battler = gBattlerAbility = gBattlerTarget;
+                    gBattlescriptCurrInstr = BattleScript_DarkTypePreventsPrankster;
+                    effect = 1;
+                }
+            #endif
+			gBattleStruct->atkCancellerTracker++;
+			break;
```
Comment (ultima-soul) [inline src/battle_util.c:3206]:
```suggestion
            gBattleStruct->atkCancellerTracker++;
            break;
```

File: include/constants/battle_config.h [hunk 4]
```diff
@@ -122,6 +122,7 @@
 #define B_FLASH_FIRE_FROZEN         GEN_6 // In Gen5+, Flash Fire can trigger even when frozen, when it couldn't before.
 #define B_SYNCHRONIZE_NATURE        GEN_6 // In Gen8+, if the Pokémon with Synchronize is leading the party, it's 100% guaranteed that wild Pokémon will have the same ability, as opposed to 50% previously.
 #define B_UPDATED_INTIMIDATE        GEN_8 // In Gen8, Intimidate doesn't work on opponents with the Inner Focus, Scrappy, Own Tempo or Oblivious abilities.
+#define B_PRANKSTER_DARK_TYPES      GEN_7 // In Gen7, Prankster-elevated status moves do not affect Dark type Pokémon.
```
Comment (ultima-soul) [inline include/constants/battle_config.h:125]:
```suggestion
#define B_PRANKSTER_DARK_TYPES      GEN_7 // In Gen7+, Prankster-elevated status moves do not affect Dark-type Pokémon.
```

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Some issues found while testing:
```
- Perish Song and Rototiller shouldn't bypass foe Dark-type Prankster immunity unlike other multi-target moves that do
- Assist should do animation, say what move is being used, and then fail due to Dark-type Prankster immunity and it should be able to choose either opponent for single target attacks rather than opponent's first slot always.
    - Also if the move chosen is something that targets ally or self, it should work regardless of type. e.g. Harden should work regardless of foes' typings.
- If Dark-type that normally would be immune to Prankster'd move has Magic Bounce ability, it bounces the move back if the move would normally be bounced back by Magic Bounce; right now it says the move failed 
```

**Comment by ghoulslash** [comment]
ready for re-review. Rototiller is not handled due to the presence of #1657. Preferably, this should be merged and then that PR can handle rototiller + prankster. Otherwise, we will need to wait for #1657

**Review by ultima-soul — CHANGES_REQUESTED** [review]
A couple of issues:
```
- Dark-type Prankster immunity of ally is bypassed for any boosted moves (e.g. Thunder Wave used on your Dark-type ally is successful)
    - This includes Perish Song

- Assist always fails with "It doesn't affect Foe X..." if slot 1 foe is Dark-type even if move should be successful like when using Curse when it should work normally

- Under the test case of both foes being Dark-type and Assist user's partner has Curse, Pound, and Razor Leaf it should behave like:
    - Curse can be invoked and be successful
    - Pound can be invoked against either foe and will print "It doesn't affect Foe X..."
    - Razor Leaf can be invoked and will print "It doesn't affect Foe X..." and "It doesn't affect Foe Y..."

- Under the test case of slot 1 foe being Dark-type and slot 3 foe being non-Dark-type and Assist user's partner has Curse, Pound, and Razor Leaf it should behave like:
    - Curse can be invoked and be successful
    - Pound can be invoked against Dark-type foe and will print "It doesn't affect Foe X..."
    - Pound can be invoked against non-Dark-type foe and be successful
    - Razor Leaf can be invoked and will print "It doesn't affect Foe X..." for slot 1 foe, do Razor Leaf animation, and then hit non-Dark-type (slot 3 foe).

- Under the test case of slot 3 foe being Dark-type and slot 1 foe being non-Dark-type and Assist user's partner has Curse, Pound, and Razor Leaf it should behave like:
    - Curse can be invoked and be successful
    - Pound can be invoked against Dark-type foe and will print "It doesn't affect Foe X..."
    - Pound can be invoked against non-Dark-type foe and be successful
    - Razor Leaf can be invoked and will print "It doesn't affect Foe X..." for slot 3 foe, do Razor Leaf animation, and then hit non-Dark-type (slot 1 foe).

- Magic Coat shouldn't automatically fail if slot 1 foe is Dark-type

- Dark-type foe that uses status move against user (type doesn't matter) of Prankster'd Magic Coat shouldn't be affected by reflected status move
    - Works like normal with bounced move back message, but then prints "It doesn't affect Foe X..."
    - If Dark-type ally uses status move against Prankster'd Magic Coat user, it bounces back successfully
```

**Comment by ghoulslash** [comment]
Ready for re-review.

I created a ProtectStruct field, `pranksterElevated`, which is set upon prankster boosting priority to handle assist and other calling-move effects. I also handle magic coat in Cmd_attackcanceler. A general function, `BlocksPrankster`, is added to help duplicate logic.

**Review by ultima-soul — CHANGES_REQUESTED** [review]
I'm not sure if this one thing can be fixed and whether it's relevant to the PR, but it is a behavioural difference from what I tested in Gen 7:
```
- Under the test case of slot 3 foe being Dark-type and slot 1 foe being non-Dark-type and Assist user's partner has Curse, Pound, and Razor Leaf it should behave like:
    - Razor Leaf can be invoked and will print "It doesn't affect Foe X..." for slot 3 foe, do Razor Leaf animation, and then hit non-Dark-type (slot 1 foe).
 ```
 Other issues are:
 ```
 - Magic Coat should bounce back all status effects for the turn and not just once

- Dark-type foe that uses status move against user (type doesn't matter) of Prankster'd Magic Coat should be affected by reflected status move if `B_PRANKSTER_DARK_TYPES` is set to Gen 6 or below

- If `B_PRANKSTER_DARK_TYPES` is set to Gen 6 or below, when Perish Song is used, it shouldn't say that Dark-type foes are not affected
    - Currently it prints this, but Perish Song is still afflicted; the affliction is correct
```

For the Magic Coat blocking issue when the config is set to Gen 6 or below, see my potential suggestion in `BlocksPrankster`.

File: asm/macros/battle_script.inc [hunk 5]
```diff
@@ -1796,6 +1796,11 @@
 	.4byte \terrainFlags
 	.4byte \ptr
 	.endm
+
```
Comment (ultima-soul) [inline asm/macros/battle_script.inc:1799]:
```suggestion

```

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Good otherwise.

**Review by ultima-soul — APPROVED** [review]

File: asm/macros/battle_script.inc [hunk 6]
```diff
@@ -1796,6 +1796,10 @@
 	.4byte \terrainFlags
 	.4byte \ptr
 	.endm
+	.macro jumpifpranksterblocked battler:req, ptr:req
```
Comment (ultima-soul) [inline asm/macros/battle_script.inc:1799]:
```suggestion
	.endm

	.macro jumpifpranksterblocked battler:req, ptr:req
```

**Review by ultima-soul — CHANGES_REQUESTED** [review]
Not sure why GitHub suggestions deleted the line.

**Review by ultima-soul — APPROVED** [review]

**Pull request merged**